### ISSUE TITLE
Corrupted AST when binding an option and accessing a property

### ISSUE THREAD

**Issue by alfonsogarciacaro** [fsharp/FSharp.Compiler.Service#531]
Hi there! I'm getting unexpected results in the AST when having a binding to an option an accessing a property immediately after. So, if I have this code:
```
let test() =
    let x = Some 5
    x.IsSome
```

The resolved expresión for `test()` body becomes:
```
Application
  (Call
  (Some Value val x,member get_IsSome,[type Microsoft.FSharp.Core.int],[],[]),
   [],[Const (null,type Microsoft.FSharp.Core.unit)])
```

The binding disappeared and has been replaced by an application to unit. This seems to happen with `.IsSome`, `.IsNone`, `Option.isSome`, `Option.isNone`. If I don't access the identifier directly everything works normally. Modifying the code like so:
```
let test() =
    let xs = Some 5
    xs |> Option.isSome
```

gives:
```
Let
  ((val xs,
    NewUnionCase
  (type Microsoft.FSharp.Core.Option<Microsoft.FSharp.Core.int>,Some,
   [Const (5,type Microsoft.FSharp.Core.int)])),
   Call
  (null,val op_PipeRight,[],
   [type Microsoft.FSharp.Core.int Microsoft.FSharp.Core.option;
    type Microsoft.FSharp.Core.bool],
   [Value val xs;
    Lambda
  (val option,
   Call (null,val isSome,[],[type Microsoft.FSharp.Core.int],[Value val option]))]))
```

I reproduced the issue with versions 2.0.0.3 and 2.0.0.5. Any hint to solve the problem would be much appreciated.

**Comment by dsyme** [comment]
Definitely a bug!

**Comment by alfonsogarciacaro** [comment]
I've found a similar issue. When I have this code:
```
type Todo = { title: string; completed: bool }

let filters =
    Map [ "active", Array.filter (fun (t: Todo) -> not t.completed) ]

let remaining todos =
    filters.["active"] todos |> Seq.length
```

The generated AST for method `remaining` is:
```
Application
  (Call
  (Some Value val filters,member get_Item,
   [type Microsoft.FSharp.Core.string;
    type Test.Todo Microsoft.FSharp.Core.[] -> Test.Todo Microsoft.FSharp.Core.[]],
   [],[Const ("active",type Microsoft.FSharp.Core.string)]),
   [],[Value val todos])
```

This means the call to `Seq.length` got missing. I'm very puzzled about this, if I change the body of the method to `Seq.length(filters.["active"] todos)` then the AST is generated correctly:

```
Call
  (null,val length,[],[type Test.Todo],
   [Coerce
  (type Microsoft.FSharp.Collections.seq<Test.Todo>,
   Application
  (Call
  (Some Value val filters,member get_Item,
   [type Microsoft.FSharp.Core.string;
    type Test.Todo Microsoft.FSharp.Core.[] -> Test.Todo Microsoft.FSharp.Core.[]],
   [],[Const ("active",type Microsoft.FSharp.Core.string)]),
   [],[Value val todos]))])
```

Also if I use a dictionary instead of a map for `filters` the problem disappears:

```
type Todo = { title: string; completed: bool }

let filters =
    dict [ "active", Array.filter (fun (t: Todo) -> not t.completed) ]

let remaining todos =
    filters.["active"] todos |> Seq.length
```

This generates the following (correct) AST for `remaining`:

```
Call
  (null,val op_PipeRight,[],
   [type Test.Todo Microsoft.FSharp.Core.[]; type Microsoft.FSharp.Core.int],
   [Application
  (Call
  (Some Value val filters,member get_Item,
   [type Microsoft.FSharp.Core.string;
    type Test.Todo Microsoft.FSharp.Core.[] -> Test.Todo Microsoft.FSharp.Core.[]],
   [],[Const ("active",type Microsoft.FSharp.Core.string)]),
   [],[Value val todos]);
    Lambda
  (val source,
   Call
  (null,val length,[],[type Test.Todo],
   [Coerce (type Microsoft.FSharp.Collections.seq<Test.Todo>,Value val source)]))])
```

Any help with this would be appreciated. I haven't found many cases so far but if the AST provided by the F# compiler services is not reliable, it will greatly hinder the usability of a project like [Fable](https://github.com/fsprojects/Fable).

**Comment by dsyme** [comment]
Well that needs to be fixed :)

**Comment by alfonsogarciacaro** [comment]
I'd like to help but I'll need some guidance about where to look at first, as I don't know the inner guts of the compiler services that well.

**Comment by dsyme** [comment]
This was fixed in 2.0.0.7 too (all the fixes were the same bug more or less!)
